### ISSUE TITLE
Hetzner Load Balancer still gets created when using use_klipper_lb = "true"

### ISSUE THREAD

**Issue by andi0b** [kube-hetzner/terraform-hcloud-kube-hetzner#245]
I'm trying to create a deployment without a Hetzner Cloud Load Balancer. The setting `use_klipper_lb = "true"` seems to work well, and the k8s LoadBalancer gets the all the nodes IPs assigned, instead of a LoadBalancer IP.

But there are still Hetzner Cloud Load Balancers created. I see them in the Hetzner Console and I get charged for them.

Any idea how to fix that?

**Comment by mysticaltech** [comment]
@andi0b Please post the out of `kubectl get svc -A -o wide` here. It could be something else requesting it from the Hetzner CCM.

**Comment by andi0b** [comment]
Do I misunderstand the functionality of `use_klipper_lb`? I expect it to replace the Hetzner LB and expose a Service of type LoadBalancer to all cluster nodes IPs. So I can point a round-robin DNS record to it and do some random/basic load balancing, without paying for a Hetzner LB.

```
❯ kubectl get nodes -o wide
NAME                     STATUS   ROLES                       AGE     VERSION        INTERNAL-IP    EXTERNAL-IP    OS-IMAGE           KERNEL-VERSION     CONTAINER-RUNTIME
control-plane-nbg1-gjd   Ready    control-plane,etcd,master   4m52s   v1.23.8+k3s2   10.254.0.102   162.55.57.0    openSUSE MicroOS   5.18.9-2-default   containerd://1.5.13-k3s1
control-plane-nbg1-swr   Ready    control-plane,etcd,master   5m25s   v1.23.8+k3s2   10.254.0.101   162.55.56.19   openSUSE MicroOS   5.18.9-2-default   containerd://1.5.13-k3s1

❯ kubectl get services -o wide -A
NAMESPACE         NAME                            TYPE           CLUSTER-IP      EXTERNAL-IP                PORT(S)                      AGE     SELECTOR
cert-manager      cert-manager                    ClusterIP      10.43.183.227   <none>                     9402/TCP                     84s     app.kubernetes.io/component=controller,app.kubernetes.io/instance=cert-manager,app.kubernetes.io/name=cert-manager
cert-manager      cert-manager-webhook            ClusterIP      10.43.216.62    <none>                     443/TCP                      84s     app.kubernetes.io/component=webhook,app.kubernetes.io/instance=cert-manager,app.kubernetes.io/name=webhook
default           kubernetes                      ClusterIP      10.43.0.1       <none>                     443/TCP                      2m45s   <none>
kube-system       hcloud-csi-controller-metrics   ClusterIP      10.43.135.39    <none>                     9189/TCP                     2m2s    app=hcloud-csi-controller
kube-system       hcloud-csi-node-metrics         ClusterIP      10.43.173.141   <none>                     9189/TCP                     2m2s    app=hcloud-csi
kube-system       kube-dns                        ClusterIP      10.43.0.10      <none>                     53/UDP,53/TCP,9153/TCP       2m41s   k8s-app=kube-dns
kube-system       metrics-server                  ClusterIP      10.43.137.143   <none>                     443/TCP                      2m41s   k8s-app=metrics-server
kube-system       traefik                         LoadBalancer   10.43.202.138   162.55.56.19,162.55.57.0   80:30052/TCP,443:31228/TCP   68s     app.kubernetes.io/instance=traefik,app.kubernetes.io/name=traefik
longhorn-system   csi-attacher                    ClusterIP      10.43.141.103   <none>                     12345/TCP                    43s     app=csi-attacher
longhorn-system   csi-provisioner                 ClusterIP      10.43.107.227   <none>                     12345/TCP                    43s     app=csi-provisioner
longhorn-system   csi-resizer                     ClusterIP      10.43.129.7     <none>                     12345/TCP                    43s     app=csi-resizer
longhorn-system   csi-snapshotter                 ClusterIP      10.43.148.234   <none>                     12345/TCP                    42s     app=csi-snapshotter
longhorn-system   longhorn-backend                ClusterIP      10.43.126.240   <none>                     9500/TCP                     82s     app=longhorn-manager
longhorn-system   longhorn-engine-manager         ClusterIP      None            <none>                     <none>                       82s     longhorn.io/component=instance-manager,longhorn.io/instance-manager-type=engine
longhorn-system   longhorn-frontend               ClusterIP      10.43.130.192   <none>                     80/TCP                       82s     app=longhorn-ui
longhorn-system   longhorn-replica-manager        ClusterIP      None            <none>                     <none>                       82s     longhorn.io/component=instance-manager,longhorn.io/instance-manager-type=replica
```

Freshly created with the (I think) most current versions. And that's what I get inside Hetzner Cloud:

![grafik](https://user-images.githubusercontent.com/2432095/178951154-f6378cc0-08cd-407f-9a11-bd394dc8d98b.png)

**Comment by mysticaltech** [comment]
@andi0b No, you got it right; this must have been a relic from an old deployment! Sometimes, it hangs at deleting the LB when you destroy the cluster, and if you force it, you have to manually destroy the LB with the hcloud cli (for ease) or the UI of course. This is out of our control. It's explained in the readme, in the takedown section here https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner#takedown

**Comment by andi0b** [comment]
No, it’s no relic from an old deployment. If I create a new service of type load balancer, it immediately creates a load balancer in Hetzner Cloud. So it’s triggered by kubernetes and not the terraform script. 

I will give it a closer look later, to find out why it happens.

**Comment by mysticaltech** [comment]
Ok, I'll have a look too with a test deployment!

**Comment by mysticaltech** [comment]
The thing is, it has to show up somewhere in the services. Otherwise, this cluster did not create it - unless I am mistaken!

**Comment by andi0b** [comment]
@mysticaltech: I think I found the problem. The Hetzner CCM has a setting `HCLOUD_LOAD_BALANCERS_ENABLED` [see docs](https://github.com/hetznercloud/hcloud-cloud-controller-manager/blob/f7d624e83c2c3475c5606306214814250922cb8a/docs/load_balancers.md#cluster-wide-defaults) that needs to be set to `false`.

I'm quite new to Terraform, but it seems to me that this environment variable is nowhere set, which means the Hetzner CCM is still creating load balancers. When I set it manually via `kubectl -n kube-system edit deployment hcloud-cloud-controller-manager`, it stops creating load balancers.

For now this is a feasible workaround for me, as you only need to set it once.

**Comment by mysticaltech** [comment]
Great find indeed. @phaer We already have use_klipper_lb to work with, so we can set it to false if that is true. 

@andi0b  There are already variables defined in the file above, so you could copy-paste on a new line and just change the name and set the conditional logic; it's not sorcery and a great way to start with Terraform! All that to say, PR is welcome! :) If not, we will do it ourselves as soon as we have time.

**Comment by andi0b** [comment]
Okay, I'm on it, will submit a pull request today if I'm successful.

**Comment by bcr-ch** [comment]
I had the same problem and it turns out there was a deployment that kept recreating the LB. Named hetzner-csi… I deleted the deployment and it stopped getting recreated.

**Comment by bcr-ch** [comment]
@andi0b sorry, you are correct.  It was the hcloud-cloud-controller-manager.  I would go in and delete the LB and then the next day it would be recreated.  I backed up the deployment before I delete it.  

```
creationTimestamp: "2022-07-01T07:12:28Z"
  generation: 1
  name: hcloud-cloud-controller-manager
  namespace: kube-system
  resourceVersion: "1219114"
  uid: c4544c02-0413-4ff2-a226-ea5fcbfdc4cc
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 2
  selector:
    matchLabels:
      app: hcloud-cloud-controller-manager
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      annotations:
        scheduler.alpha.kubernetes.io/critical-pod: ""
      creationTimestamp: null
      labels:
        app: hcloud-cloud-controller-manager
    spec:
      containers:
      - command:
        - /bin/hcloud-cloud-controller-manager
        - --cloud-provider=hcloud
        - --leader-elect=false
        - --allow-untagged-cloud
        - --allocate-node-cidrs=true
        - --cluster-cidr=10.42.0.0/16
        env:
        - name: HCLOUD_LOAD_BALANCERS_LOCATION
          value: fsn1
        - name: HCLOUD_LOAD_BALANCERS_USE_PRIVATE_IP
          value: "true"
        - name: NODE_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: spec.nodeName
```

**Comment by andi0b** [comment]
@bcr-ch exactly that we are fixing right now. You can reinstall that service with the additional env-variable ´HCLOUD_LOAD_BALANCERS_ENABLED´ set to false

**Comment by mysticaltech** [comment]
Fix by @andi0b #246 🥳 has been merged to staging and will ship in the next release later today.

**Comment by bcr-ch** [comment]
> @bcr-ch exactly that we are fixing right now. You can reinstall that service with the additional env-variable `HCLOUD_LOAD_BALANCERS_ENABLED` set to false

Done as you suggested and so far no load balancers have been created.  Added the environment variable to my backup of the deployment and redeployed with kubectl apply -f backup_lb.yaml

Thanks again

**Comment by mysticaltech** [comment]
Released in v1.4.2.
